### DISCUSSION TITLE
How to choose the best possible version on reported Vulnerabilities

### DISCUSSION THREAD

**Discussion by sundarv85** [aquasecurity/trivy#7219] [Q&A]
### Question

Hello,

We are facing an issue where for one package there are mulitple versions, and when we use one of the most fitting versions, then we get new vulnerabilities reported. 

For e.g.

We are currently using Django version 2.0.5 and have performed a vulnerability scan, which produced the following results:

#### Initial Scan Results:
```
"VulnerabilityID": "CVE-2020-13254",
"FixedVersion": "2.2.13, 3.0.7",

"VulnerabilityID": "CVE-2021-33203",
"FixedVersion": "2.2.24, 3.1.12, 3.2.4",

"VulnerabilityID": "CVE-2021-44420",
"FixedVersion": "2.2.25, 3.1.14, 3.2.10",

"VulnerabilityID": "CVE-2019-3498",
"FixedVersion": "1.11.18, 2.0.10, 2.1.5",

"VulnerabilityID": "CVE-2024-24680",
"FixedVersion": "3.2.24, 4.2.10, 5.0.2",
```

Based on these results, I upgraded to the least necessary fixed version that addresses all the CVEs, 3.2.24. However, running another scan with Trivy on version 3.2.24 showed the following new vulnerabilities:

#### New Scan Results with Django 3.2.24:
```
"VulnerabilityID": "CVE-2024-27351",
"FixedVersion": "3.2.25, 4.2.11, 5.0.3",
```

It seems that upgrading to `3.2.24` has introduced new vulnerabilities (same happens with 5.0.2 as well). I am looking for guidance on how to choose the correct version that resolves all the identified vulnerabilities without introducing new ones.

#### Specific Questions:

1. Is there a recommended strategy for determining the best upgrade path in such cases where multiple fixed versions are listed?
2. How can I ensure that the version I choose addresses all the known vulnerabilities without introducing new ones?

Any help or advice on this would be greatly appreciated.

Thank you!

### Target

Filesystem

### Scanner

Vulnerability

### Output Format

JSON

### Mode

Standalone

### Operating System

Ubuntu

### Version

```bash
v0.53.0
```

**Comment by knqyf263** [comment]
Thanks for sharing your problem. We're aware of this demand, but Trivy currently doesn't support it.
It may be possible to detect a clean version of that particular package, but it currently seems difficult to show a clean version for the whole project, as upgrading one package will affect the versions of other packages. In addition, complex options may be required, such as whether major version upgrades are acceptable.

**Reply by sundarv85** [reply]
Thanks, @knqyf263.

Understood. Do you have any suggestions on how to accomplish that, assuming major version upgrades are acceptable? If you have any, it would be great to hear them, and we can see if we can implement them ourselves for our use case.

One option would be to pass back the highest acceptable range to Trivy and check if there are no known vulnerabilities.

Any other suggestions would be most welcome.

**Reply by knqyf263** [reply]
It is not a very ideal way to do this, but it could be possible to generate an SBOM and have Trivy scan it, then change the version and rescan it to see the results.
